### PR TITLE
KREST-578: Add configurable DosFilter support.

### PR DESCRIPTION
**Pull request by rigelbm**
This PR exposes per-application DosFilter support. See https://www.eclipse.org/jetty/javadoc/jetty-9/org/eclipse/jetty/servlets/DoSFilter.html for more information about the options available.

Config names are kept equal to the ones in Jetty's DosFilter. That unfortunately includes a usage of "whitelist", which otherwise would not meet Confluent's bar.

### CODE REVIEW

**Review by dimitarndimitrov — APPROVED** [review]
+1.
In the context of our discussion today:
- it's not great that we have to settle for something that most likely won't really guard us very well;
- an upper-level-of-abstraction solution, like maybe Envoy, should definitely take this off our hands, and do a better job guarding us;

File: core/src/main/java/io/confluent/rest/RestConfig.java [hunk 1]
```diff
@@ -796,4 +950,56 @@ private static void validateHeaderConfigAction(String action) {
               + "The action need be one of [\"set\", \"add\", \"setDate\", \"addDate\"]", action));
     }
   }
+
+  public final boolean getDosFilterEnabled() {
+    return getBoolean(DOS_FILTER_ENABLED_CONFIG);
+  }
```
[the remark this answers is not in the document]
Comment (rigelbm) [reply core/src/main/java/io/confluent/rest/RestConfig.java]: Done.

File: core/src/main/java/io/confluent/rest/RestConfig.java [hunk 2]
```diff
@@ -344,6 +346,80 @@
   protected static final String CSRF_PREVENTION_TOKEN_MAX_ENTRIES_DOC =
       "Specifies the maximum number of entries the token cache may contain";
 
+  private static final String DOS_FILTER_ENABLED_CONFIG = "dos.filter.enabled";
```
[the remark this answers is not in the document]
Comment (rigelbm) [reply core/src/main/java/io/confluent/rest/RestConfig.java:349]: Acknowledged.

I think the `_CONFIG` prefix helps differentiates the config name from the actual config value. Ideally, the config name, default and doc should be wrapped in a single object.

File: core/src/main/java/io/confluent/rest/RestConfig.java [hunk 3]
```diff
@@ -344,6 +346,80 @@
...
+  private static final String DOS_FILTER_ENABLED_DOC =
+      "Whether to enable DosFilter for the application.";
```
[the remark this answers is not in the document]
Comment (rigelbm) [reply core/src/main/java/io/confluent/rest/RestConfig.java]: Done.

File: core/src/main/java/io/confluent/rest/RestConfig.java [hunk 4]
```diff
@@ -344,6 +346,80 @@
...
+  private static final String DOS_FILTER_IP_WHITELIST_CONFIG = "dos.filter.ip.whitelist";
+  private static final String DOS_FILTER_IP_WHITELIST_DOC =
+      "A comma-separated list of IP addresses that will not be rate limited.";
+  private static final List<String> DOS_FILTER_IP_WHITELIST_DEFAULT = emptyList();
+
+  private static final String DOS_FILTER_MANAGED_ATTR_CONFIG = "dos.filter.managed.attr";
+  private static final String DOS_FILTER_MANAGED_ATTR_DOC =
+      "If set to true, then this servlet is set as a ServletContext attribute with the filter "
+          + "name as the attribute name. This allows a context external mechanism (for example, "
+          + "JMX via ContextHandler.MANAGED_ATTRIBUTES) to manage the configuration of the filter.";
```
[the remark this answers is not in the document]
Comment (rigelbm) [reply core/src/main/java/io/confluent/rest/RestConfig.java]: Done.

File: core/src/test/java/io/confluent/rest/DosFilterTest.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
+package io.confluent.rest;
```
[the remark this answers is not in the document]
Comment (rigelbm) [reply core/src/test/java/io/confluent/rest/DosFilterTest.java:17]: Done.

File: core/src/test/java/io/confluent/rest/DosFilterTest.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+    Client client = ClientBuilder.newClient();
+
+    // Request should succeed.
+    Response response1 = client.target(server.getURI()).path("/foo").request().get();
+    assertEquals(Status.OK.getStatusCode(), response1.getStatus());
+
+    // Request should be throttled.
+    Response response2 = client.target(server.getURI()).path("/foo").request().get();
+    assertEquals(Status.TOO_MANY_REQUESTS.getStatusCode(), response2.getStatus());
```
[the remark this answers is not in the document]
Comment (rigelbm) [reply core/src/test/java/io/confluent/rest/DosFilterTest.java]: Done.

**Pull request merged**